processing/src/main/java/io/druid/query/groupby/GroupByQuery.java

```diff
@@ -281,10 +290,93 @@ public int compare(Object lhs, Object rhs)
     );
   }
 
-  public Ordering<Row> getRowOrdering(final boolean granular)
+  private Ordering<Row> getRowOrderingForPushDown(
+      final boolean granular,
+      final DefaultLimitSpec limitSpec,
+      final List<AggregatorFactory> aggregatorSpecs
+  )
   {
     final boolean sortByDimsFirst = getContextSortByDimsFirst();
 
+    final List<String> orderedFieldNames = new ArrayList<>();
+    final Set<Integer> dimsInOrderBy = new HashSet<>();
+    final List<Integer> directions = new ArrayList<>();
+
+    for (OrderByColumnSpec orderSpec : limitSpec.getColumns()) {
+      int direction = orderSpec.getDirection() == OrderByColumnSpec.Direction.ASCENDING ? 1 : -1;
+      int dimIndex = OrderByColumnSpec.getDimIndexForOrderBy(orderSpec, dimensions);
+      if (dimIndex >= 0) {
+        DimensionSpec dim = dimensions.get(dimIndex);
+        orderedFieldNames.add(dim.getOutputName());
+        dimsInOrderBy.add(dimIndex);
+        directions.add(direction);
+      }
+    }
+
+    for (int i = 0; i < dimensions.size(); i++) {
+      if (!dimsInOrderBy.contains(i)) {
+        orderedFieldNames.add(dimensions.get(i).getOutputName());
+        directions.add(1);
+      }
+    }
+
+    final Comparator<Row> timeComparator = getTimeComparator(granular);
+
+    if (timeComparator == null) {
+      return Ordering.from(
+          new Comparator<Row>()
+          {
+            @Override
+            public int compare(Row lhs, Row rhs)
+            {
+              return compareDimsAndAggs(orderedFieldNames, directions, lhs, rhs);
+            }
+          }
+      );
+    } else if (sortByDimsFirst) {
+      return Ordering.from(
+          new Comparator<Row>()
+          {
+            @Override
+            public int compare(Row lhs, Row rhs)
+            {
+              final int cmp = compareDimsAndAggs(orderedFieldNames, directions, lhs, rhs);
+              if (cmp != 0) {
+                return cmp;
+              }
+
+              return timeComparator.compare(lhs, rhs);
+            }
+          }
+      );
+    } else {
+      return Ordering.from(
+          new Comparator<Row>()
+          {
+            @Override
+            public int compare(Row lhs, Row rhs)
+            {
+              final int timeCompare = timeComparator.compare(lhs, rhs);
+
+              if (timeCompare != 0) {
+                return timeCompare;
+              }
+
+              return compareDimsAndAggs(orderedFieldNames, directions, lhs, rhs);
+            }
+          }
+      );
+    }
+  }
+
+  public Ordering<Row> getRowOrdering(final boolean granular)
+  {
+    final boolean pushDownSort = getContextBoolean(GroupByQueryConfig.CTX_KEY_PUSH_DOWN_LIMIT, false);
+    if (pushDownSort && !DefaultLimitSpec.sortingOrderHasNonGroupingFields(limitSpec, dimensions)) {
+      return getRowOrderingForPushDown(granular, (DefaultLimitSpec) limitSpec, aggregatorSpecs);
+    }
```

Comment:
so , here it is expected limits are pushed only when pushDownSort is explicitly set by the user in the request.. while in RowBasedGrouperHelper it will happen automatically if limit was non-negative and ordering columns were subset of dimensions, wouldn't this produce wrong results if user did not enable pushDown but it was automatic.... am i missing something?
also, "whether limit is pushed down or not" code can be refactored into a common place and is used here and RowBasedGrouperHelper so that no disparity can happen.